File: PhysPanelLib/KeyboardManager.cs
// Xbox Fullscreen Experience Tool
// Copyright (C) 2025 8bit2qubit

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System.Diagnostics;
using System.Runtime.InteropServices;

namespace PhysPanelLib
{
    /// <summary>
    /// 表示找不到 TabTip.exe 執行檔時發生的錯誤。
    /// </summary>
    public class TabTipNotFoundException : Exception
    {
        // 提供一個標準的建構函式
        public TabTipNotFoundException(string message) : base(message) { }
    }

    /// <summary>
    /// 表示啟動或透過 COM 啟用觸控鍵盤失敗時發生的錯誤。
    /// </summary>
    public class TabTipActivationException : Exception
    {
        // 提供一個標準的建構函式，可以包含內部錯誤訊息
        public TabTipActivationException(string message, Exception? innerException) : base(message, innerException) { }
    }

    /// <summary>
    /// 提供與 Windows 觸控鍵盤 (TabTip.exe) 互動的功能。
    /// </summary>
    public static class KeyboardManager
    {
        #region Win32 & COM Definitions
        [ComImport, Guid("4CE576FA-83DC-4F88-951C-9D0782B4E376")]
        private class TipInvocation { }

        [ComImport, Guid("37c994e7-432b-4834-a2f7-dce1f13b834b"), InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
        private interface ITipInvocation { void Toggle(IntPtr hwnd); }
        #endregion

        private const string TabTipProcessName = "TabTip";
        private const string TabTipFileName = "TabTip.exe";
        private const string ShellProcessName = "explorer";
        private static readonly TimeSpan ShellReadyTimeout = TimeSpan.FromSeconds(30);
        private static readonly TimeSpan ComServiceTimeout = TimeSpan.FromSeconds(10);

        // 在啟動鍵盤後加入的延遲，給予足夠的反應時間。
        private const int PostLaunchDelayMs = 5000;

        private static readonly Lazy<string?> TabTipPath = new Lazy<string?>(() =>
        {
            string commonProgramFiles = Environment.GetFolderPath(Environment.SpecialFolder.CommonProgramFiles);
            string path = Path.Combine(commonProgramFiles, @"Microsoft Shared\ink", TabTipFileName);
            return File.Exists(path) ? path : null;
        });

        /// <summary>
        /// 確保觸控鍵盤處於「收起」狀態，並使其處理程序在背景待命。
        /// 此方法包含等待 explorer.exe 就緒的邏輯，適用於系統登入時執行。
        /// </summary>
        public static void StartTouchKeyboard()
        {
            // 如果處理程序已在執行，直接返回
            if (Process.GetProcessesByName(TabTipProcessName).Any())
            {
                return; // 處理程序已在執行，直接返回
            }

            // 如果找不到 TabTip.exe 的路徑，記錄錯誤並返回
            if (string.IsNullOrEmpty(TabTipPath.Value))
            {
                // 拋出一個明確的錯誤，而不是靜默失敗
                throw new TabTipNotFoundException($"{TabTipFileName} executable not found in its expected path.");
            }

            try
            {
                // 步驟 1: 等待 explorer.exe 準備就緒
                if (!WaitForExplorerProcess(ShellReadyTimeout))
                {
                    throw new TabTipActivationException($"Timed out waiting for the Windows Shell ({ShellProcessName}.exe) to start.", null);
                }

                // 步驟 2: 啟動 TabTip.exe 處理程序
                var startInfo = new ProcessStartInfo(TabTipPath.Value) { UseShellExecute = true };
                Process.Start(startInfo);

                // 步驟 3: 等待固定時間，讓系統的桌面環境完全準備好
                Thread.Sleep(PostLaunchDelayMs);

                // 步驟 4: 輪詢 COM 服務，直到它準備就緒
                ITipInvocation? tipInvocation = PollForComService(ComServiceTimeout);

                // 步驟 5: 如果成功取得 COM 物件，發送命令將其隱藏
                if (tipInvocation != null)
                {
                    try
                    {
                        tipInvocation.Toggle(IntPtr.Zero); // 執行 "切換" 來隱藏觸控鍵盤
                    }
                    finally
                    {
                        Marshal.ReleaseComObject(tipInvocation);
                    }
                }
                else
                {
                    throw new TabTipActivationException($"Timed out waiting for the {TabTipFileName} COM service to become available.", null);
                }
            }
            catch (Exception ex) when (ex is not TabTipActivationException && ex is not TabTipNotFoundException)
            {
                throw new TabTipActivationException($"An unexpected error occurred while starting or hiding {TabTipFileName}.", ex);
            }
        }

        /// <summary>
        /// 等待 explorer.exe 處理程序啟動。
        /// 這是確保桌面環境已準備好接收 COM 命令的關鍵步驟。
        /// </summary>
        /// <param name="timeout">最長等待時間。</param>
        /// <returns>如果處理程序在超時前出現，則為 true；否則為 false。</returns>
        private static bool WaitForExplorerProcess(TimeSpan timeout)
        {
            var stopwatch = Stopwatch.StartNew();
            while (stopwatch.Elapsed < timeout)
            {
                if (Process.GetProcessesByName(ShellProcessName).Any())
                {
                    return true; // 找到了
                }
                Thread.Sleep(500); // 每半秒檢查一次
            }
            return false; // 超時
        }

        /// <summary>
        /// 在指定的時間內，持續輪詢 COM 服務，嘗試建立 TipInvocation 物件。
        /// 這是為了應對處理程序啟動後 COM 服務需要時間初始化的情況。
        /// </summary>
        /// <returns>成功時返回 ITipInvocation 物件，逾時則返回 null。</returns>
        private static ITipInvocation? PollForComService(TimeSpan timeout)
        {
            var stopwatch = Stopwatch.StartNew();
            while (stopwatch.Elapsed < timeout)
            {
                try
                {
                    return (ITipInvocation)new TipInvocation();
                }
                catch (COMException)
                {
                    Thread.Sleep(250); // 等待一小段時間再重試
                }
            }
            return null;
        }
    }
}

File: PhysPanelCPP/KeyboardManager.cpp
// Xbox Full Screen Experience Tool
// Copyright (C) 2025 8bit2qubit

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include "pch.h"
#include "KeyboardManager.h"

std::string WstringToString(const std::wstring& wstr) {
    if (wstr.empty()) return std::string();
    int size_needed = WideCharToMultiByte(CP_UTF8, 0, &wstr[0], (int)wstr.size(), NULL, 0, NULL, NULL);
    std::string strTo(size_needed, 0);
    WideCharToMultiByte(CP_UTF8, 0, &wstr[0], (int)wstr.size(), &strTo[0], size_needed, NULL, NULL);
    return strTo;
}

namespace KeyboardManager {

    void LogDebug(const wchar_t* format, ...) {
#if defined(_DEBUG)
        va_list args;
        va_start(args, format);
        vwprintf(format, args);
        wprintf(L"\n");
        va_end(args);
#else
        UNREFERENCED_PARAMETER(format);
#endif
    }

    struct __declspec(uuid("4CE576FA-83DC-4F88-951C-9D0782B4E376")) TipInvocation;
    struct __declspec(uuid("37C994E7-432B-4834-A2F7-DCE1F13B834B")) ITipInvocation : IUnknown {
        virtual HRESULT __stdcall Toggle(HWND hwnd) = 0;
    };

    struct __declspec(uuid("D5120AA3-46BA-44C5-822D-CA8092C1FC72")) FrameworkInputPane;
    struct __declspec(uuid("5752238B-24F0-495A-82F1-2FD593056796")) IFrameworkInputPane : IUnknown {
        virtual HRESULT __stdcall Advise(IUnknown* pWindow, IUnknown* pHandler, DWORD* pdwCookie) = 0;
        virtual HRESULT __stdcall AdviseWithHWND(HWND hwnd, IUnknown* pHandler, DWORD* pdwCookie) = 0;
        virtual HRESULT __stdcall Unadvise(DWORD pdwCookie) = 0;
        virtual HRESULT __stdcall Location(RECT* prcInputPaneScreenLocation) = 0;
    };

    constexpr auto TABTIP_PROCESS_NAME = L"TabTip.exe";
    constexpr auto SHELL_PROCESS_NAME = L"explorer.exe";
    const auto SHELL_READY_TIMEOUT = std::chrono::seconds(30);
    const auto COM_SERVICE_TIMEOUT = std::chrono::seconds(10);
    //const auto POST_LAUNCH_DELAY = std::chrono::seconds(0);

    bool IsProcessRunning(const wchar_t* processName) {
        HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
        if (snapshot == INVALID_HANDLE_VALUE) return false;
        PROCESSENTRY32W entry;
        entry.dwSize = sizeof(entry);
        bool found = false;
        if (Process32FirstW(snapshot, &entry)) {
            do {
                if (_wcsicmp(entry.szExeFile, processName) == 0) {
                    found = true;
                    break;
                }
            } while (Process32NextW(snapshot, &entry));
        }
        CloseHandle(snapshot);
        return found;
    }

    bool WaitForProcess(const wchar_t* processName, std::chrono::milliseconds timeout) {
        LogDebug(L"[Debug] Waiting for process: %s", processName);
        auto start = std::chrono::high_resolution_clock::now();
        while (std::chrono::high_resolution_clock::now() - start < timeout) {
            if (IsProcessRunning(processName)) {
                LogDebug(L"[Debug] Process found: %s", processName);
                return true;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(500));
        }
        LogDebug(L"[Debug] Wait for process TIMEOUT: %s", processName);
        return false;
    }

    bool IsTouchKeyboardVisible() {
        bool isVisible = false;
        IFrameworkInputPane* pInputPane = nullptr;

        HRESULT hr = CoCreateInstance(__uuidof(FrameworkInputPane), NULL, CLSCTX_INPROC_SERVER, __uuidof(IFrameworkInputPane), (void**)&pInputPane);

        if (SUCCEEDED(hr) && pInputPane) {
            RECT rc = { 0 };
            hr = pInputPane->Location(&rc);

            if (SUCCEEDED(hr)) {
                if ((rc.right - rc.left) > 0 && (rc.bottom - rc.top) > 0) {
                    isVisible = true;
                    LogDebug(L"[Debug] IFrameworkInputPane State: Visible (Rect: %d x %d)", (rc.right - rc.left), (rc.bottom - rc.top));
                }
            }
            pInputPane->Release();
        }

        return isVisible;
    }

    void StartTouchKeyboard() {
        LogDebug(L"--- StartTouchKeyboard() [Mode: Start -> Poll -> Hide] ---");

        if (!IsProcessRunning(TABTIP_PROCESS_NAME)) {
            PWSTR pszPath = nullptr;
            HRESULT hr_path = SHGetKnownFolderPath(FOLDERID_ProgramFilesCommon, 0, NULL, &pszPath);
            if (FAILED(hr_path)) {
                LogDebug(L"[Debug] Error: Failed to retrieve FOLDERID_ProgramFilesCommon.");
                throw TabTipNotFoundException("Failed to retrieve Common Program Files path.");
            }

            std::wstring tabTipPath(pszPath);
            CoTaskMemFree(pszPath);
            tabTipPath += L"\\Microsoft Shared\\ink\\TabTip.exe";
            LogDebug(L"[Debug] TabTip path: %s", tabTipPath.c_str());

            DWORD fileAttr = GetFileAttributesW(tabTipPath.c_str());
            if (fileAttr == INVALID_FILE_ATTRIBUTES || (fileAttr & FILE_ATTRIBUTE_DIRECTORY)) {
                LogDebug(L"[Debug] Warning: TabTip.exe not found at expected path.");
                throw TabTipNotFoundException("TabTip.exe not found at its expected path.");
            }

            if (!WaitForProcess(SHELL_PROCESS_NAME, SHELL_READY_TIMEOUT)) {
                throw TabTipActivationException("Timed out waiting for Windows Shell (explorer.exe).");
            }

            LogDebug(L"[Debug] Launching TabTip.exe service via ShellExecuteW...");
            ShellExecuteW(NULL, L"open", tabTipPath.c_str(), NULL, NULL, SW_SHOWNORMAL);

            WaitForProcess(TABTIP_PROCESS_NAME, std::chrono::seconds(10));

            //LogDebug(L"[Debug] Starting post-launch delay...");
            //std::this_thread::sleep_for(POST_LAUNCH_DELAY);
            //LogDebug(L"[Debug] Post-launch delay finished.");
        } else {
            LogDebug(L"[Debug] TabTip.exe is already running. Skipping launch, proceeding to visibility check.");
        }

        bool comInitialized = false;
        try {
            LogDebug(L"[Debug] Initializing COM for visibility polling and control...");
            HRESULT hr = CoInitializeEx(NULL, COINIT_APARTMENTTHREADED);
            if (SUCCEEDED(hr)) {
                comInitialized = true;
            }

            LogDebug(L"[Debug] Starting visibility poll (max 10s)...");
            bool visible = false;
            auto pollStart = std::chrono::high_resolution_clock::now();

            while (std::chrono::high_resolution_clock::now() - pollStart < COM_SERVICE_TIMEOUT) {
                if (IsTouchKeyboardVisible()) {
                    visible = true;
                    LogDebug(L"[Debug] Poll Result: Keyboard is VISIBLE.");
                    break;
                }
                std::this_thread::sleep_for(std::chrono::milliseconds(250));
            }

            if (visible) {
                LogDebug(L"[Debug] Action: Keyboard visible. Toggling to HIDE.");

                ITipInvocation* pTip = nullptr;
                auto start = std::chrono::high_resolution_clock::now();

                while (std::chrono::high_resolution_clock::now() - start < COM_SERVICE_TIMEOUT) {
                    hr = CoCreateInstance(__uuidof(TipInvocation), NULL, CLSCTX_LOCAL_SERVER, __uuidof(ITipInvocation), (void**)&pTip);
                    if (SUCCEEDED(hr)) break;
                    std::this_thread::sleep_for(std::chrono::milliseconds(250));
                }

                if (pTip) {
                    LogDebug(L"[Debug] COM service connected. Invoking Toggle() to HIDE keyboard.");
                    pTip->Toggle(GetDesktopWindow());
                    pTip->Release();
                    LogDebug(L"[Debug] Keyboard hidden successfully.");
                }
                else {
                    LogDebug(L"[Debug] FAILED to connect to COM service (pTip is null). Cannot HIDE keyboard.");
                    throw TabTipActivationException("Failed to connect to TabTip COM interface (Keyboard detected but unresponsive).");
                }
            }
            else {
                LogDebug(L"[Debug] Poll Completed: Keyboard never appeared. Assuming silent background execution.");
            }

            if (comInitialized) {
                LogDebug(L"[Debug] CoUninitialize.");
                CoUninitialize();
            }
        }
        catch (const _com_error& e) {
            if (comInitialized) CoUninitialize();
            throw TabTipActivationException(WstringToString(e.ErrorMessage()));
        }
        catch (const TabTipActivationException& e) {
            UNREFERENCED_PARAMETER(e);
            if (comInitialized) CoUninitialize();
            throw;
        }
        catch (...) {
            LogDebug(L"[Debug] Critical: Unknown EXCEPTION caught.");
            if (comInitialized) CoUninitialize();
            throw;
        }

        LogDebug(L"--- StartTouchKeyboard() Finished ---");
    }
}